### PR TITLE
Add `Tydepef<Tag, T>` template

### PR DESCRIPTION
**Pull request by akopich**
To promote type-safety we introduce a type-safe wrapper
generalizing the ideas behind `Baseline` and `Comparison`
wrappers, which are only type aliases now.

This also introduces a function `Call` that allows to
write the client code in wrapper-agnostic way, yet
make a call on wrapped arguments, preserving all the type
checks.

Tests: Compile, Unit, Manual
Bug: http://b/236715501

### CODE REVIEW

File: src/OrbitBase/include/OrbitBase/Typedef.h [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+using EnableIfUConvertibleToT = std::enable_if_t<std::is_convertible_v<U, T>>;
+
+// Strong typedef
+template <typename Tag, typename T>
+class Typedef {
+ public:
+  const T* operator->() const { return &value_; }
+  T* operator->() { return &value_; }
+
+  const T& operator*() const& { return value_; }
```
Comment (beckerhe) [inline src/OrbitBase/include/OrbitBase/Typedef.h:24]: You should also define the last combination `const&&`

File: src/OrbitBase/include/OrbitBase/Typedef.h [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+  T&& operator*() && { return std::move(value_); }
+  T& operator*() & { return value_; }
```
Comment (beckerhe) [inline src/OrbitBase/include/OrbitBase/Typedef.h:26]: All the operators should be marked nodiscard

File: src/OrbitBase/TypedefTest.cpp [hunk 3]
```diff
@@ -0,0 +1,88 @@
...
+
+}  // namespace
+
+static int Add(int i, int j) { return i + j; }
+
+namespace orbit_base {
+template <typename T>
+using Wrapper = Typedef<TestTag, T>;
+
+TEST(TypedefTest, CanInstantiate) {
```
Comment (beckerhe) [inline src/OrbitBase/TypedefTest.cpp:29]: You should also add a test with a move-only-type (std::unique_ptr<int>) and something that's neither copyable nor movable (I think std::mutex should fit that)

File: src/OrbitBase/include/OrbitBase/Typedef.h [hunk 4]
```diff
@@ -0,0 +1,64 @@
...
+#include <utility>
+
+namespace orbit_base {
+
+template <typename T, typename U>
+using EnableIfUConvertibleToT = std::enable_if_t<std::is_convertible_v<U, T>>;
+
+// Strong typedef
+template <typename Tag, typename T>
+class Typedef {
```
Comment (beckerhe) [inline src/OrbitBase/include/OrbitBase/Typedef.h:19]: I think you can mark every function in that class `constexpr`

File: src/OrbitBase/include/OrbitBase/Typedef.h [hunk 5]
```diff
@@ -0,0 +1,64 @@
...
+ public:
+  const T* operator->() const { return &value_; }
+  T* operator->() { return &value_; }
+
+  const T& operator*() const& { return value_; }
+  T&& operator*() && { return std::move(value_); }
+  T& operator*() & { return value_; }
+
+  template <typename U, typename = EnableIfUConvertibleToT<T, U>>
+  explicit Typedef(U&& value) : value_(std::forward<U>(value)) {}
```
Comment (beckerhe) [inline src/OrbitBase/include/OrbitBase/Typedef.h:29]: Can you make `explicit` depend on whether `U` can be converted implicitly or explicitly to `T`?

This probably needs to overloads that are enabled/disable via SFINAE

File: src/OrbitBase/TypedefTest.cpp [hunk 6]
```diff
@@ -0,0 +1,88 @@
...
+  EXPECT_EQ(*wrapper_of_non_const, non_const);
+
+  Wrapper<int> wrapper_of_literal(1);
+  EXPECT_EQ(*wrapper_of_literal, 1);
+
+  Wrapper<std::string> wrapper_of_string_literal("foo");
+  EXPECT_EQ(*wrapper_of_string_literal, "foo");
+}
+
+TEST(TypedefTest, CallIsCorrect) {
```
Comment (beckerhe) [inline src/OrbitBase/TypedefTest.cpp:49]: This should also be tested with move-only types.

File: src/OrbitBase/TypedefTest.cpp [hunk 7]
```diff
@@ -0,0 +1,88 @@
...
+  }
+
+  {
+    const Wrapper<int> sum_wrapped = Call(AddStruct::Add, kFirstWrapped, kSecondWrapped);
+    EXPECT_EQ(*sum_wrapped, kSum);
+  }
+
+  {
+    bool was_called = false;
+    auto returns_void = [&was_called](int /*i*/) { was_called = true; };
```
Comment (beckerhe) [inline src/OrbitBase/TypedefTest.cpp:81]: Nit: You could also check whether the argument `i` is correct.

File: src/OrbitBase/TypedefTest.cpp [hunk 8]
```diff
@@ -0,0 +1,88 @@
...
+    EXPECT_EQ(*sum_wrapped, kSum);
+  }
+
+  {
+    const Wrapper<int> sum_wrapped = Call(Add, kFirstWrapped, kSecondWrapped);
+    EXPECT_EQ(*sum_wrapped, kSum);
+  }
+
+  {
+    const Wrapper<int> sum_wrapped = Call(AddStruct::Add, kFirstWrapped, kSecondWrapped);
```
Comment (beckerhe) [inline src/OrbitBase/TypedefTest.cpp:75]: Not sure what `AddStruct` is about. You're calling a static function which should not behave differently to a  free function. Do you wanted to test a member function call? Then you need to do something like this:

```cpp
struct Integer {
  int value;
  
  Integer Add(const Integer& other) const { return Integer{value + other.value}; }
};
```

File: src/OrbitBase/CMakeLists.txt [hunk 9]
```diff
@@ -123,28 +123,29 @@ target_sources(OrbitBaseTests PRIVATE
...
 # Threadpool test contains some sleeps we couldn't work around - disable them on the CI
 # since they are flaky there (but they work fine on local workstations).
-if (NOT (WIN32 AND "$ENV{QT_QPA_PLATFORM}" STREQUAL "offscreen"))
-target_sources(OrbitBaseTests PRIVATE
-        ThreadPoolTest.cpp
-)
+if(NOT(WIN32 AND "$ENV{QT_QPA_PLATFORM}" STREQUAL "offscreen"))
+        target_sources(OrbitBaseTests PRIVATE
+                ThreadPoolTest.cpp
+        )
```
Comment (beckerhe) [inline src/OrbitBase/CMakeLists.txt:148]: These are unrelated changes and shouldn't be part of the PR

File: src/OrbitBase/CMakeLists.txt [hunk 10]
```diff
@@ -71,30 +72,29 @@ target_sources(OrbitBase PRIVATE
...
+                GetProcessIdsLinux.cpp
+                ThreadUtilsLinux.cpp)
 endif()
 
 target_link_libraries(OrbitBase PUBLIC
         CONAN_PKG::abseil
         CONAN_PKG::outcome
         std::filesystem)
 
-
```
Comment (beckerhe) [inline src/OrbitBase/CMakeLists.txt:97]: These are unrelated changes and shouldn't be part of the PR

**Review by beckerhe — COMMENTED** [review]
I haven't looked at the Mizar changes yet. Only the OrbitBase part

File: src/OrbitBase/TypedefTest.cpp (hunk 8 above)
Comment (akopich) [reply src/OrbitBase/TypedefTest.cpp:75]: > calling a static function which should not behave differently to a free function
NVM then. I'll remove it.

File: src/OrbitBase/CMakeLists.txt (hunk 10 above)
Comment (akopich) [reply src/OrbitBase/CMakeLists.txt]: Argh. Sorry. Autoformat.

File: src/OrbitBase/CMakeLists.txt (hunk 9 above)
Comment (akopich) [reply src/OrbitBase/CMakeLists.txt]: Argh. Sorry. Autoformat.

File: src/OrbitBase/TypedefTest.cpp (hunk 8 above)
Comment (akopich) [reply src/OrbitBase/TypedefTest.cpp]: Done.

File: src/OrbitBase/include/OrbitBase/Typedef.h [hunk 11]
```diff
@@ -0,0 +1,121 @@
...
+  constexpr Typedef(const Typedef<Tag, U>& other) : value_(other.value_) {}
+
+  template <typename U, typename = EnableIfUConvertibleToT<T, U>>
+  constexpr Typedef(Typedef<Tag, U>&& other) : value_(std::move(other.value_)) {}
+
+  template <typename U, typename = EnableIfUIsNotConvertibleToT<T, U>, typename = void>
+  constexpr explicit Typedef(const Typedef<Tag, U>& other) : value_(other.value_) {}
+
+  template <typename U, typename = EnableIfUIsNotConvertibleToT<T, U>, typename = void>
+  constexpr explicit Typedef(Typedef<Tag, U>&& other) : value_(std::move(other.value_)) {}
```
Comment (beckerhe) [inline src/OrbitBase/include/OrbitBase/Typedef.h:54]: I don't understand the additional `typename = void` parameter. You said without it it didn't work. Have you figured out why? I would love to understand if that's really fixing a potential problem or just silencing the symptom.

File: src/OrbitBase/TypedefTest.cpp [hunk 12]
```diff
@@ -0,0 +1,230 @@
...
+  EXPECT_EQ(*wrapper_of_constexpr, kConstexprInt);
+
+  int non_const = 1;
+  Wrapper<int> wrapper_of_non_const(non_const);
+  EXPECT_EQ(*wrapper_of_non_const, non_const);
+
+  Wrapper<int> wrapper_of_literal(1);
+  EXPECT_EQ(*wrapper_of_literal, 1);
+
+  Wrapper<std::string> wrapper_of_string_literal("foo");
```
Comment (beckerhe) [inline src/OrbitBase/TypedefTest.cpp:64]: Nit: Once it's wrapped it's not a string literal anymore. It's just a string which was initialized from a literal

File: src/OrbitBase/TypedefTest.cpp [hunk 13]
```diff
@@ -0,0 +1,230 @@
...
+
+struct D {
+  int value{};
+  explicit D(A&& a) : value(std::move(a.value)) {}
+  D(const A& a) = delete;
+};
+
+}  // namespace
+
+static int Add(int i, int j) { return i + j; }
```
Comment (beckerhe) [inline src/OrbitBase/TypedefTest.cpp:42]: Is that function used anywhere?

File: src/OrbitBase/TypedefTest.cpp [hunk 14]
```diff
@@ -0,0 +1,230 @@
...
+  }
+
+  {
+    auto add = [](int&& i, int&& j) { return i + j; };
+
+    Wrapper<int> first(kFirst);
+    Wrapper<int> second(kSecond);
+    const Wrapper<int> sum_wrapped = Apply(add, std::move(first), std::move(second));
+    EXPECT_EQ(*sum_wrapped, kSum);
+  }
```
Comment (beckerhe) [inline src/OrbitBase/TypedefTest.cpp:188]: Could you also add a test where only the second parameter is moved to ensure that perfect forwarding also works if it's different for the parameters?

File: src/MizarBase/include/MizarBase/BaselineOrComparison.h [hunk 15]
```diff
@@ -0,0 +1,37 @@
...
+
+template <typename T, typename... Args>
+Baseline<T> MakeBaseline(Args&&... args) {
+  return Baseline<T>(T(std::forward<Args>(args)...));
+}
+
+template <typename T, typename... Args>
+Comparison<T> MakeComparison(Args&&... args) {
+  return Comparison<T>(T(std::forward<Args>(args)...));
+}
```
Comment (beckerhe) [inline src/MizarBase/include/MizarBase/BaselineOrComparison.h:33]: Nit: nodiscard

File: src/MizarBase/include/MizarBase/BaselineOrComparison.h [hunk 16]
```diff
@@ -0,0 +1,37 @@
+// Copyright (c) 2022 The Orbit Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+#ifndef MIZAR_BASE_BASELINE_OR_COMPARISON_H_
+#define MIZAR_BASE_BASELINE_OR_COMPARISON_H_
+
+#include <functional>
+#include <type_traits>
```
Comment (beckerhe) [inline src/MizarBase/include/MizarBase/BaselineOrComparison.h:9]: I think these are not needed, right?

File: src/OrbitBase/include/OrbitBase/Typedef.h [hunk 17]
```diff
@@ -0,0 +1,121 @@
...
+
+namespace orbit_base {
+
+template <typename T, typename U>
+using EnableIfUConvertibleToT = std::enable_if_t<std::is_convertible_v<U, T>>;
+
+template <typename T, typename U>
+using EnableIfUIsNotConvertibleToT = std::enable_if_t<!std::is_convertible_v<U, T>>;
+
+// Strong typedef
```
Comment (beckerhe) [inline src/OrbitBase/include/OrbitBase/Typedef.h:20]: A little more elaborate explanation would be great. :-) Maybe even a small example on how to use that with `Apply` for the most basic use-case.

File: src/OrbitBase/include/OrbitBase/Typedef.h (hunk 11 above)
Comment (akopich) [reply src/OrbitBase/include/OrbitBase/Typedef.h:54]: > You said without it it didn't work. Have you figured out why? 
I think, the compiler generates an error at the first phase of name lookup. There it does see a redefinition of constructor that differs only by constexpr (and you can't overload by constexpr) and by the default template value, which starts to matter at the second phase.

**Review by florian-kuebler — COMMENTED** [review]
As said in the other PR. I am not really sure if this is going into the right direction, but now you did the work. On the usage (Mizar) side, the code became more readable than before this change, which is good.

I have to admit, that I was not able to get anything in the Typedef.h, so I focused on the tests and the usage. That looks good from my side.

I'd like @dpallotti and @beckerhe to decide here.

File: src/OrbitBase/TypedefTest.cpp (hunk 13 above)
Comment (akopich) [reply src/OrbitBase/TypedefTest.cpp:42]: Yeah, in one of the tests.

File: src/OrbitBase/include/OrbitBase/Typedef.h [hunk 18]
```diff
@@ -0,0 +1,121 @@
...
+struct AreSame : std::true_type {};
+
+template <typename T, typename U, typename... TT>
+struct AreSame<T, U, TT...>
+    : std::integral_constant<bool, std::is_same<T, U>{} && AreSame<T, TT...>{}> {};
+
+template <typename Head, typename... Tail>
+struct First {
+  using type = Head;
+};
```
[the remark this answers is not in the document]
Comment (akopich) [reply src/OrbitBase/include/OrbitBase/Typedef.h:99]: I got rid of them both.

**Comment by dpallotti** [comment]
First of all, I'll save myself the review as @beckerhe is by far the best for the job anyway. The only think I will reiterate though is: `Typedef` needs to be extensively documented and have code examples (style guide: "you should make sure that tricky code is especially well commented").

> As said in the other PR. I am not really sure if this is going into the right direction, but now you did the work. On the usage (Mizar) side, the code became more readable than before this change, which is good.

The broader discussion on the philosophy of C++. Do you want to learn templates in details and enjoy the challenge they pose and the possibilities they open? Then you probably want to start using templates everywhere. You don't, like most? Then templates quickly get absolutely unreadable.

The style guide is quite clear. Please refer to https://google.github.io/styleguide/cppguide.html#Template_metaprogramming.

Having read that, as a library tool this seems fine. The important thing with complex templates is that the complexity remains isolated, which should be the case here (style guide: "you should expect to put considerable effort into minimizing and isolating the complexity").

Should we start using this type extensively? I don't think so and there is no need, but that's a different story.

**Comment by akopich** [comment]
@dpallotti 
> Do you want to learn templates in details and enjoy the challenge they pose and the possibilities they open? 

I personally do. 

> Should we start using this type extensively?

Ideally, yes. Now we have timestamp, function_id, scope_id and probably smth else. And all of these are `uint64_t`. Now we could define smth like `using Timestamp = Typedef<TimestampTag, uint64_t>`. This would make function signatures drastically more readable and enforce typechecks. E.g. `CaptureData::GetTimersForScope` takes three uint64_t as of now.  OTOH, I understand this can't be done overnight, as we extensively use protos in the parts of the code that do not handle serialisation (and we shouldn't). But that's a different story.

**Comment by akopich** [comment]
@dpallotti 

> you should make sure that tricky code is especially well commented

Are you satisfied with the comment above the definition of Typedef ?

**Comment by dpallotti** [comment]
> I personally do.

Trust me, this was clear.

> Now we have timestamp, function_id, scope_id and probably smth else. And all of these are uint64_t. Now we could define smth like using Timestamp = Typedef<TimestampTag, uint64_t>. This would make function signatures drastically more readable and enforce typechecks.

Fair. What's the advantage over just using a `struct` containing a single value? Easier interface/conversions/etc.?

File: src/OrbitBase/include/OrbitBase/Typedef.h [hunk 19]
```diff
@@ -0,0 +1,127 @@
...
+// ```
+// template <typename T>
+// using Wrapper = Typedef<TestTag, T>;
+// ```
+// Now we can instantiate the typedefs
+// ```
+// const Wrapper<int> kFirstWrapped(1);
+// const Wrapper<int> kSecondWrapped(2);
+// ```
+// and define a callable (not necessarily a lambda)
```
Comment (dpallotti) [inline src/OrbitBase/include/OrbitBase/Typedef.h:34]: I think there's a bit of a conceptual jump here. It sounds like the only point of `Typedef` is `Apply`. It should be one paragraph describing `Typedef` and one describing `Apply`.

File: src/OrbitBase/include/OrbitBase/Typedef.h [hunk 20]
```diff
@@ -0,0 +1,127 @@
...
+  }
+
+ private:
+  T value_;
+};
+
+template <typename Tag>
+class Typedef<Tag, void> {};
+
+// `action` is a callable. `args` are its args wrapped in `Typedef`s  with the same `Tag`. The args
```
Comment (dpallotti) [inline src/OrbitBase/include/OrbitBase/Typedef.h:105]: > its args wrapped in `Typedef`s with the same `Tag`

This seems arbitrary. If two `Typedef`s wrap the same type, why would it not be allowed to call a function on them? Either `Apply` is too generic as a name (I think this is the case here), or the choice is arbitrary for a specific use-case.

**Comment by dpallotti** [comment]
> Or do you mean, we should also comment on the implementation, not only the usage ?

I meant the usage, it's way more important. But if you have comments to add for the implementation you are welcome to do that.

**Comment by akopich** [comment]
> Fair. What's the advantage over just using a struct containing a single value? Easier interface/conversions/etc.?

That's one thing. Another thing, `Apply` helps you to easily "stay in container". It allows you to write code agnostic of the wrappers and relying only on wrappees. Then you could it to Apply along with the wrappers, have the typechecks (e.g. that you don't add time and ids) and you have your return value properly wrapped too. See how it's used in Mizar for example.

**Comment by dpallotti** [comment]
> Another thing, `Apply` helps you to easily "stay in container". It allows you to write code agnostic of the wrappers and relying only on wrappees. Then you could it to Apply along with the wrappers, have the typechecks (e.g. that you don't add time and ids) and you have your return value properly wrapped too. See how it's used in Mizar for example.

I think that in many cases this is not what you want. If a function works an a `TimestampNs` type, it should take `TimestampNs`.

File: src/OrbitBase/include/OrbitBase/Typedef.h (hunk 19 above)
Comment (akopich) [reply src/OrbitBase/include/OrbitBase/Typedef.h:34]: I'll just move to the comment above `Apply`

File: src/OrbitBase/include/OrbitBase/Typedef.h (hunk 20 above)
Comment (akopich) [reply src/OrbitBase/include/OrbitBase/Typedef.h:105]: I don't think it's arbitrary. You don't want smth like `Time total = Apply(sum, duration1, scope_id)` to compile. 
What actually happens here: we _lift_ the function into container Timestamp<Tag, T> and apply it. Perhaps. `LiftAndApply` would be a better name, but _lift_ is FP terminology.

Comment (akopich) [reply src/OrbitBase/include/OrbitBase/Typedef.h:105]: Moreover, I DO want to allow for calling a function on Typedefs of the same Tag, yet of different T. This is already used in Mizar.

Comment (dpallotti) [reply src/OrbitBase/include/OrbitBase/Typedef.h:105]: In relation to the name, it's arbitrary. You could have a function that takes a duration and a scope id. We need a better name.

Comment (akopich) [reply src/OrbitBase/include/OrbitBase/Typedef.h:105]: Are you OK with LiftAndApply?

Comment (dpallotti) [reply src/OrbitBase/include/OrbitBase/Typedef.h:105]: Yeah, I can't think of anything better.

**Comment by dpallotti** [comment]
> > I think that in many cases this is not what you want. If a function works an a TimestampNs type, it should take TimestampNs.
> 
> Sometimes, maybe. Sometimes you would even want to wrap the result in a different Typdef. E.g. `Elapsed GetElapsed(Timestamp, Timestamp)`.
> 
> But imagine, you wish to have a similar behaviour for differently wrapped types. E.g. you want to be able to get max of `Timestamp`s and `Duration`s. Without apply you'd have to implement the same logic (leave alone wrapping/unwrapping) twice. Apply allows you to just use the same `int max(int, int)` in both cases! Again, this is already utilised in Mizar where I do the same stuff on objects wrapped in Comparison and Baseline.

I think `Comparison` and `Baseline` are a bit too ad hoc for the example (you added `Apply` for those, right?), but I see this can be useful. Thanks for explaining.

**Comment by akopich** [comment]
> you added Apply for those, right? 

Sort of. At first, I had just two wrappers `Baseline<T>` and `Comparison<T>` and I didn't want to implement some pieces of logic twice. So the implementation of that logic had to be parameterised with a template template argument. Reviewers didn't like it and it was decided to move the heavily templated stuff to a library for the sake of localisation. Then, naturally, some generalisation happened.

File: src/OrbitBase/include/OrbitBase/Typedef.h (hunk 20 above)
Comment (akopich) [reply src/OrbitBase/include/OrbitBase/Typedef.h:105]: @beckerhe are you OK with `LiftAndApply` ?

Comment (beckerhe) [reply src/OrbitBase/include/OrbitBase/Typedef.h]: yes
